### PR TITLE
Moved remote client out of docker backend

### PR DESCRIPTION
**Pull request by kpavel**
- to make it more generic moved ibm_vpc client files from pywren_ibm_cloud/libs/docker/clients
to pywren_ibm_cloud/libs/clients
- changed directory where docker remote files stored from default /temp to /pywren-temp
- removed remote_client calls to start/stop remote system on every usage
- added remote_client.start if needed to the compute.py constructor
- added compute.cleanup to stop remote system using remote_client

- to avoid compute.cleanup called on executor garbage collection automatically specify pywren.data_cleaner : False in the config file and call it on demand using executor_instance.clean().
 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

### CODE REVIEW

File: pywren_ibm_cloud/executor.py [hunk 1]
```diff
@@ -461,6 +460,8 @@ def clean(self, fs=None, cs=None, cloudobjects=True, force=True, log=True):
         Deletes all the files from COS. These files include the function,
         the data serialization and the function invocation results.
         """
+        self.invoker.cleanup()
```
Comment (gilv) [inline pywren_ibm_cloud/executor.py:463]: @kpavel this step deletes all data from COS. in many flows, we will need to stop VM but keep data in COS

**Comment by JosepSampe** [comment]
In my opinion there is no need to modify `compute.py`, `invoker.py` and `executor.py`.  compute.py is a very high level abstraction of all compute backends, and the logic added here seems to be very tied to the docker backend. Is there any reason to not to add the logic you added in `compute.py` into the docker.py `__init__` method? 

Also, as @gilv mentioned, I think the `clean()` method is not the correct place to "stop" the VM. You can simply add the `__del__` method into the `docker.py` and then the VM will be automatically stopped once the Compute instance is deleted. This way there is no need to add extra logic tied in the `invoker.py` and `executor.py` files since everything would be automatic. You can also use the `atexit` module as we do in the localhost backend to stop the VM automatically when python exits.

**Comment by gilv** [comment]
@JosepSampe sometimes we have
pw.docker(func1,....)
pw.docker(func2,..)

so we don't want to stop VM between them. Based on your suggestion if we use __del__ method of the docker.py then VM will be stoped between steps

**Comment by erezhdd** [comment]
> In my opinion there is no need to modify `compute.py`, `invoker.py` and `executor.py`. `compute.py` is a very high level abstraction of all compute backends, and the logic added here seems to be very tied to the docker backend. Is there any reason to not to add the logic you added in `compute.py` into the docker.py `__init__` method?
> 
Actually, the whole point of this design is separating the compute backend from the initialization/finalization to avoid too many sub-types of back-ends who differ only in the environment being initialized or destroyed, such as docker_vpc, docker_ecs, ow_vpc, ow_ec2, localhost_ec2, etc. With this design, compute back-ends remain unchanged, but can be combined with various initialization and finalization options. This contributes to long-term code sustainability, by maintaining fewer independent components. compute.py is the level just above the compute backend and focuses on managing low-level compute, so it seems logical to implement the template method pattern there.

**Comment by gilv** [comment]
@JosepSampe any input?

**Comment by JosepSampe** [comment]
@erezh16 Thanks for clarifying this. So the idea here is that some of the executors (not only docker) can be backed by a VM, right?

@gilv I think that with the current code it is not possible to run multiple maps from the same executor (if you get the results from each map), since the VM is stopped after getting the results, for example the next example will fail after the first get_result():

```python
pw = pywren.docker_executor()
pw.map(fn1)
res1 = pw.get_result() << -- VM is stopped here
<<-- from here all the next calls will fail
pw.map(fn2) 
res2 = pw.get_result()
pw.map(fn3)
res3 = pw.get_result()
...
```

So I'm wondering when the VM should be stopped...

In addition, now the VM is also stopped if a user uses the executor context manager, so the example you provided above will stop VM between steps as the "stop" signal was within the `__del__` method. But for the context manager I think it is ok.

**Comment by gilv** [comment]
@JosepSampe I think the pull request of @kpavel  will not stop VM between stages. @kpavel is this corrent?
```
pw = pywren.docker_executor()
pw.map(fn1)
res1 = pw.get_result() << -- VM is NOT stopped here with the PR of @kpavel 
<<-- from here all the next calls will fail
pw.map(fn2) 
res2 = pw.get_result()
pw.map(fn3)
res3 = pw.get_result()
```

**Comment by gilv** [comment]
@kpavel @JosepSampe just run

```
    pw = pywren.docker_executor(log_level='DEBUG', storage_backend='ibm_cos')
    futures = pw.map(my_first_map_function, iterdata)
    print (pw.get_result())
    futures = pw.map(my_second_map_function, iterdata)
    print (pw.get_result())

    pw.clean()
    pw.dismantle()
```

Perfectly works

**Comment by kpavel** [comment]
> @JosepSampe I think the pull request of @kpavel will not stop VM between stages. @kpavel is this corrent?
> 
> ```
> pw = pywren.docker_executor()
> pw.map(fn1)
> res1 = pw.get_result() << -- VM is NOT stopped here with the PR of @kpavel 
> <<-- from here all the next calls will fail
> pw.map(fn2) 
> res2 = pw.get_result()
> pw.map(fn3)
> res3 = pw.get_result()
> ```

Correct. Although, there an option to automatically trigger dismantle by setting "auto_dismantle" attribute to True in the pywren section in config.

**Comment by erezhdd** [comment]
Would it make sense to add a simple test to prove the point?

**Comment by JosepSampe** [comment]
@gilv Yep, that's the point, the `auto_dismantle` parameter is False by default, but following the docs seems that is must be set to True. So following the docs I expected the VM to be automatically stopped even if I run more than one `map()`.

**Comment by gilv** [comment]
@JosepSampe hmm...good point related  auto_dismantle..

**Comment by kpavel** [comment]
> @gilv Yep, that's the point, the `auto_dismantle` parameter is False by default, but following the docs seems that it must be set to True. So following the docs I expected the VM to be automatically stopped even if I run more than one `map()`.

You right, documentation updated

**Comment by JosepSampe** [comment]
Thanks @kpavel, This is now much more clear. 

My main concern was always how to deal with novice users that doesn't know how things internally work (I have 60+ new of them every year ;) ). For advanced users I think it is great to have the possibility to manually stop the VM when needed. But for novice users I think the best option is to set it to True by default. Why? When coding an application, it is very common that the code raises exceptions everywhere, so the `exec.dismantle()` method,  probably located at the end of the script, could be never called, leaving the VM started (for days? weeks?).

Additionally, I think it would be great that the users could run multiple calls from the same executor when the `auto_dismantle` param is set to True. So I'm wondering if the `wait()` method is the correct place to put it. Why not put it in in the `__del__` method of the Executor? This way there will be a consistent behaviour between traditional usage and context manager, and the `dismantle` method will be called even if there is an exception in the user's code.

For example, these both scripts will work correctly in the same way if "auto_dismantle" param is set to True:

```python
pw = pywren.docker_executor()
pw.map(fn1)
res1 = pw.get_result()
pw.map(fn2)
res2 = pw.get_result()

pw = pywren.docker_executor()
pw.map(fn1)
res1 = pw.get_result()
pw.map(fn2) 
res2 = pw.get_result()
```

```python
with pywren.ibm_cf_executor() as pw:
       pw.map(fn1)
       res1 = pw.get_result()
       pw.map(fn2) 
       res2 = pw.get_result()

with pywren.ibm_cf_executor() as pw:
       pw.map(fn1)
       res1 = pw.get_result()
       pw.map(fn2) 
       res2 = pw.get_result()
```

What do you think @kpavel @gilv? ot is there something that I'm missing that requires to stop the VM after getting the results?

**Comment by kpavel** [comment]
> Thanks @kpavel, This is now much more clear.
> 
> My main concern was always how to deal with novice users that doesn't know how things internally work (I have 60+ new of them every year ;) ). For advanced users I think it is great to have the possibility to manually stop the VM when needed. But for novice users I think the best option is to set it to True by default. Why? When coding an application, it is very common that the code raises exceptions everywhere, so the `exec.dismantle()` method, probably located at the end of the script, could be never called, leaving the VM started (for days? weeks?).
> 
> Additionally, I think it would be great that the users could run multiple calls from the same executor when the `auto_dismantle` param is set to True. So I'm wondering if the `wait()` method is the correct place to put it. Why not put it in in the `__del__` method of the Executor? This way there will be a consistent behaviour between traditional usage and context manager, and the `dismantle` method will be called even if there is an exception in the user's code.
> 
> For example, these both scripts will work correctly in the same way if "auto_dismantle" param is set to True:
> 
> ```python
> pw = pywren.docker_executor()
> pw.map(fn1)
> res1 = pw.get_result()
> pw.map(fn2)
> res2 = pw.get_result()
> 
> pw = pywren.docker_executor()
> pw.map(fn1)
> res1 = pw.get_result()
> pw.map(fn2) 
> res2 = pw.get_result()
> ```
> 
> ```python
> with pywren.ibm_cf_executor() as pw:
>        pw.map(fn1)
>        res1 = pw.get_result()
>        pw.map(fn2) 
>        res2 = pw.get_result()
> 
> with pywren.ibm_cf_executor() as pw:
>        pw.map(fn1)
>        res1 = pw.get_result()
>        pw.map(fn2) 
>        res2 = pw.get_result()
> ```
> 
> What do you think @kpavel @gilv? ot is there something that I'm missing that requires to stop the VM after getting the results?

@JosepSampe From my experience, you can't count on __del__ method for such things. I also checked in pywren and didn't notice that executor.__del__ or compute.__del__ called ever.

**Comment by gilv** [comment]
@JosepSampe i agree with your points...this is also my fear that something not by the "book" and then VM will not be stoped.  I prefer to be on the safe side, so i don't mind to use __del__  as you suggested. But @kpavel claims that __del__ will not solve the issue as it not always called, so seems __del__ will not solve the issue. So the question i guess when __del__ is called and in which cases it's not

**Comment by JosepSampe** [comment]
@gilv I ran some experiments and in my case both `executor.__del__` and `compute.__del__` are always called correctly, even if  I provoque an exception. @kpavel In what circumstances do you see the `__del__` method is not called?

The only constraint with adding logic in `__del__` is that this logic always works ok except when `__del__` is called when a python script has finished. During the python finalization process it is not possible to spawns process or open files, so I'm wondering if it is possible to perform a call to the VPC API...

**Comment by kpavel** [comment]
> @gilv I ran some experiments and in my case both `executor.__del__` and `compute.__del__` are always called correctly, even if I provoque an exception. @kpavel In what circumstances do you see the `__del__` method is not called?
> 
> The only constraint with adding logic in `__del__` is that this logic always works ok except when `__del__` is called when a python script has finished. During the python finalization process it is not possible to spawns process or open files, so I'm wondering if it is possible to perform a call to the VPC API...

Changed the auto_dismantle to True by default. As well added call to dismantle from `executor.__del__`
In my environment (Python 3.7.4), i never see that `executor.__del__` called. To make it more safe we should recommend users to use context manager or call the dismantle explicitly.

**Comment by gilv** [comment]
@JosepSampe any more comments?

**Comment by JosepSampe** [comment]
@JosepSampe Nop ;) If I could replicate it, I will investigate in the future why with that environment the `__del__` method is not called, and if there is any alternative

**Comment by gilv** [comment]
@JosepSampe i also plan to see who is right here - you claiming it's called or @kpavel who claims it'a not ;)...

**Comment by JosepSampe** [comment]
@gilv I think we are all correct: The `__del__` method is called when the object is destructed, i.e. if there is some reference of the executor stored somewhere, the ` __del__` method is never called. If instead all the references are deleted, then the `__del__` method is called. So depending of the script and how the user coded the application, it is possible that the `__del__` method could be never called during the execution. In any case, I think it is always called when the script finishes its execution.

**Pull request merged**